File: python/wallet/base_provider.py
"""Abstract base provider: compatible get_account_info and sign_tx.

Providers hold a keystore instance. Call `init()` to load credentials from
keystore, or use the classmethod `create()` on concrete providers.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Optional

from common.logger import get_logger
from keystore.base import KeystoreBase
from keystore.keystore import Keystore
from wallet.types import AccountInfo, SignedTxResult


class BaseProvider(ABC):
    """Abstract provider base class."""

    def __init__(
        self,
        keystore_path: Optional[str] = None,
        keystore_password: Optional[str] = None,
        keystore: Optional[KeystoreBase] = None,
    ):
        # Allow injecting a custom keystore implementation.
        self.keystore: KeystoreBase = keystore or Keystore(
            file_path=keystore_path, password=keystore_password
        )

    async def init(self) -> "BaseProvider":
        """Load keystore (constructors cannot be async)."""
        log = get_logger(__name__)
        log.debug("provider keystore init: reading keystore path=%s", self.keystore.get_path())
        self.keystore.read()
        log.debug("provider keystore init: loaded path=%s", self.keystore.get_path())
        return self

    @abstractmethod
    async def get_account_info(self) -> AccountInfo:
        """Get account info; must include wallet address."""
        ...

    @abstractmethod
    async def sign_tx(self, unsigned_tx: Any) -> SignedTxResult:
        """Sign an unsigned transaction and return signed result."""
        ...

File: typescript/src/wallet/base_provider.ts
import type { AccountInfo, SignedTxResult } from './types';
import { Keystore, KeystoreBase, type KeystoreOptions } from '../keystore';
import { logger } from '../logger';

/**
 * Abstract base provider: compatible getAccountInfo and signTx.
 * Subclasses (TronProvider, FlashProvider) implement chain-specific logic.
 *
 * Each provider holds a Keystore instance. Call `init()` to load credentials
 * from keystore, or use the static `create()` factory on concrete providers.
 */
export abstract class BaseProvider {
  /** Keystore instance for reading/writing account credentials. */
  public keystore: KeystoreBase;

  constructor(keystore?: KeystoreBase | KeystoreOptions) {
    // Allow injecting a custom keystore implementation.
    if (
      keystore &&
      typeof (keystore as any).read === 'function' &&
      typeof (keystore as any).write === 'function'
    ) {
      this.keystore = keystore as KeystoreBase;
    } else {
      this.keystore = new Keystore(keystore as KeystoreOptions | undefined);
    }
  }

  /**
   * Load credentials from keystore. Subclasses override to populate
   * chain-specific fields (privateKey, apiKey, etc.) from keystore data.
   * Must be called after construction (constructors cannot be async).
   */
  async init(): Promise<this> {
    logger.debug(
      { keystorePath: this.keystore.getPath() },
      'provider keystore init: reading keystore',
    );
    await this.keystore.read();
    logger.debug(
      { keystorePath: this.keystore.getPath() },
      'provider keystore init: loaded',
    );
    return this;
  }

  /** Get account info; must include wallet address. */
  abstract getAccountInfo(): Promise<AccountInfo>;

  /** Sign an arbitrary message (bytes) and return signature string. */
  abstract signMessage(message: Uint8Array): Promise<string>;

  /**
   * Sign an unsigned transaction (or message request) and return the signed result.
   * @param unsignedTx - Chain-specific unsigned payload
   */
  abstract signTx(unsignedTx: unknown): Promise<SignedTxResult>;
}

